_posts/2022-08-29-sbt-tutorial.md

```diff
@@ -0,0 +1,514 @@
+We can use _clean_ command to remove all the generated class files.
+
+We can run the application by using the command _run_. If we have multiple main classes in the project, sbt will show a prompt to select the main class to run. 
+In such case, we can use _runMain_ command by passing the class to run as:
+```
+runMain com.rockthejvm.Main 
+```
+
+SBT also allows to do automatic compilation on any source file change. For that, we can start the compile command prefixed by ~ symbol within the sbt shell:
+```
+~compile
+```
+Now, sbt will track the source files and if any change is detected in the files, it will automatically recompile them.
+
+## Adding external Dependenncies
+So far, we have created scala files without using any external dependencies. Now, let's look at how we can add additional libraries to our project. We can add the dependencies to out project by adding it to the settings _libraryDependencies_. SBT follows ivy style format to define the dependencies as :
+```
+libraryDependencies += "com.lihaoyi" %% "fansi" % "0.4.0"
+```
+
+The method `+=` appends the provided library to the project dependencies. 
+An sbt dependency contains mainly 3 parts separated by % symbol. 
+The first part is the groupId of the library. The second part is the library name(artifactId) and the third part is the version of the library to be used. If you notice, you can see that a double percentage symbol (%%) is used between groupId and artifactId. 
+Scala is not binary compatible with different versions (such as 2.11, 2.12, 2.13 etc). Hence there are separate releases for each scala libraries for each required versions. %% symbol ensures that sbt uses the same version of library as the project.
+That means, sbt will automatically append the scala version of the project before trying to find the library. The above dependency code is equivalent to the following format(note that single % is used, but artifactid contains the scala major version):
+```
+libraryDependencies += "com.lihaoyi" % "fansi_2.13" % "0.4.0"
+``` 
+
+SBT also has triple % symbol (%%%), which is used to resolve ScalaJS libraries.
+
+We can add multiple dependencies together by using ++= and Seq :
+```
+libraryDependencies ++= Seq(
+  "com.lihaoyi" % "fansi_2.13" % "0.4.0"
+  //Can add more depencies here.
+)
+``` 
+
+## Scala files under project directory
+SBT allows us to keep supporting files for the build as scala files. This helps to configure most of the required configurations in the known format of scala classes. We can place the scala files under the directory _project_. 
+Generally we keep the version info as an object in the scala files and these are accessible directly inside the build.sbt files. This was we can keep and track all the dependencies in a single places and access between multiple parts of the project.
```

Comment:
A rephrase:

Generally, depending on the project complexity, we keep various info (e.g. library dependencies) as `object` in these Scala files; they are accessible directly inside `build.sbt`. This is done so we can keep and track all the dependencies in a single place, and access them in multiple parts (modules) of the project (more on modules shortly).